static/app/components/hovercard.tsx

```diff
@@ -149,91 +119,160 @@ class Hovercard extends React.Component<Props, State> {
-
-    return (
-      <Manager>
-        <Reference>
-          {({ref}) => (
-            <span
-              ref={ref}
-              aria-describedby={this.tooltipId}
-              className={containerClassName}
-              {...hoverProps}
-            >
-              {this.props.children}
-            </span>
-          )}
-        </Reference>
-        {visible &&
-          (header || body) &&
-          ReactDOM.createPortal(
-            <Popper placement={position} modifiers={popperModifiers}>
-              {({ref, style, placement, arrowProps, scheduleUpdate}) => {
-                this.scheduleUpdate = scheduleUpdate;
-                return (
-                  <StyledHovercard
-                    id={this.tooltipId}
-                    visible={visible}
-                    ref={ref}
-                    style={style}
-                    placement={placement as Direction}
-                    offset={offset}
-                    className={cx}
-                    {...hoverProps}
-                  >
-                    {header && <Header>{header}</Header>}
-                    {body && <Body className={bodyClassName}>{body}</Body>}
-                    <HovercardArrow
-                      ref={arrowProps.ref}
-                      style={arrowProps.style}
-                      placement={placement as Direction}
-                      tipColor={tipColor}
-                      tipBorderColor={tipBorderColor}
-                    />
-                  </StyledHovercard>
-                );
-              }}
-            </Popper>,
-            this.portalEl
-          )}
-      </Manager>
-    );
-  }
+  const isVisible = React.useMemo(() => {
+    // If show is not set, then visibility state is uncontrolled
+    if (props.show === undefined) {
+      return visible;
+    }
+    return props.show;
+  }, [props.show, visible]);
+
+  const hoverProps = React.useMemo((): Pick<
+    React.HTMLProps<HTMLDivElement>,
+    'onMouseEnter' | 'onMouseLeave'
+  > => {
+    // If show is not set, then visibility state is controlled by mouse events
+    if (props.show === undefined) {
+      return {
+        onMouseEnter: () => toggleHovercard(true),
+        onMouseLeave: () => toggleHovercard(false),
+      };
+    }
+    return {};
+  }, [props.show, toggleHovercard]);
+
+  return (
+    <Manager>
+      <Reference>
+        {({ref}) => (
+          <span
+            ref={ref}
+            aria-describedby={tooltipId}
+            className={props.containerClassName}
+            {...hoverProps}
+          >
+            {props.children}
+          </span>
+        )}
+      </Reference>
+      {ReactDOM.createPortal(
+        <Popper placement={props.position ?? 'top'} modifiers={popperModifiers}>
+          {({ref, style, placement, arrowProps, scheduleUpdate}) => {
+            // Nothing to render
+            if (!props.body && !props.header) {
```

Comment:
should we use this condition before creating a portal so we don't render empty elements in the dom?

Reply:
<Popper/> doesnt actually renders anything here, iirc it just instantiates the positioning engine. Usually with things like these, I like to render them eagerly because as they mount, they tend to compute some data about their position, infer window size or container etc. If we render that at the same time as we want our card to render, we may introduce some extra work for the engine to do before it can render our card (frame drops or janky UX) whereas by keeping it like this, it can just focus on rendering the card with the data that it has already inferred when it mounted.

The counter case for that is that we now have a component taking up memory when it wouldn't have needed to, but I think we should try and optimize for rendering the hovercard asap as a priority

Reply:
oh, that's super interesting! thanks for the great explanation 🙏